test/e2e/petset.go

```diff
@@ -171,6 +198,119 @@ var _ = framework.KubeDescribe("PetSet [Slow] [Feature:PetSet]", func() {
+			podTracker = []podEvent{}
+			_, controller = cache.NewInformer(
+				&cache.ListWatch{
+					ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+						options.LabelSelector = labelSelector
+						return f.Client.Pods(ns).List(options)
+					},
+					WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+						options.LabelSelector = labelSelector
+						return f.Client.Pods(ns).Watch(options)
+					},
+				},
+				&api.Pod{},
+				0,
+				cache.ResourceEventHandlerFuncs{
+					AddFunc: func(obj interface{}) {
+						podTracker = append(podTracker, podEvent{
+							pod:   obj.(*api.Pod),
+							event: add,
+						})
+					},
+					DeleteFunc: func(obj interface{}) {
+						podTracker = append(podTracker, podEvent{
+							pod:   obj.(*api.Pod),
+							event: del,
+						})
+					},
+				},
+			)
+			go controller.Run(stopCh)
+		})
+
+		AfterEach(func() {
+			if CurrentGinkgoTestDescription().Failed {
+				dumpDebugInfo(c, ns)
+			}
+			framework.Logf("Deleting all petset in ns %v", ns)
+			deleteAllPetSets(c, ns)
+			close(stopCh)
+		})
+
+		It("Scale up/down in predictable order", func() {
+			By("creating petset " + psName + " in namespace " + ns)
+			ps := newPetSet(psName, ns, headlessSvcName, 1, nil, nil, labels)
+			ps, err := c.Apps().PetSets(ns).Create(ps)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("waiting until all petset " + psName + " replicas will be running in namespace " + ns)
+			pst := &petSetTester{c: c}
+			pst.waitForRunning(ps.Spec.Replicas, ps)
+
+			By("confirming that petset scale up will halt with unhealthy pet")
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.PetSet) { ps.Spec.Replicas = 3 })
+			pst.confirmPetCount(1, ps, 10*time.Second)
+
+			By("scalling up petset " + psName + " to 3 replicas and waiting until all of them will be running in namespace " + ns)
+			pst.setHealthy(ps)
+			pst.waitForRunning(2, ps)
+			pst.setHealthy(ps)
+			pst.waitForRunning(3, ps)
+
+			By("verifying that petset " + psName + " was scaled up in order")
+			expectedOrder := []string{"pet-0", "pet-1", "pet-2"}
+			for i, event := range Added(podTracker) {
+				if event.pod.Name != expectedOrder[i] {
+					framework.Failf("Order of scale up is wrong: %v != %v", expectedOrder[i], event.pod.Name)
+				}
+			}
+
+			By("scale down will halt with unhealthy pet")
+			pst.setUnhealthy(ps)
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.PetSet) { ps.Spec.Replicas = 0 })
+			pst.confirmPetCount(3, ps, 10*time.Second)
+
+			By("scalling down petset " + psName + " to 0 replicas and waiting until none of pods will run " + ns)
+			pst.setHealthy(ps)
+			pst.scale(ps, 0)
+
+			By("verifying that petset " + psName + " was scaled down in reverse order")
+			expectedOrder = []string{"pet-2", "pet-1", "pet-0"}
+			for i, event := range Deleted(podTracker) {
+				if event.pod.Name != expectedOrder[i] {
+					framework.Failf("Order of scale down is wrong: %v != %v", expectedOrder[i], event.pod.Name)
+				}
```

Comment:
can you please add one that does a scale up, scale down, scale up (without waiting for anything inbetween), and checks that we converge to the final replica count in order? (i.e going from 3->5, 5->2, 2->4 should not ever create pets 4,3 in that order)